src/storage.ts

```diff
@@ -73,39 +74,16 @@ export class Storage {
   // fetchBinary returns the path to a Coder binary.
   // The binary will be cached if a matching server version already exists.
   public async fetchBinary(): Promise<string | undefined> {
+    await this.cleanUpOldBinaries()
     const baseURL = this.getURL()
     if (!baseURL) {
       throw new Error("Must be logged in!")
     }
     const baseURI = vscode.Uri.parse(baseURL)
 
     const buildInfo = await getBuildInfo()
-    const binPath = this.binaryPath(buildInfo.version)
-    const exists = await fs
-      .stat(binPath)
-      .then(() => true)
-      .catch(() => false)
-    if (exists) {
-      // Even if the file exists, it could be corrupted.
-      // We run `coder version` to ensure the binary can be executed.
-      this.output.appendLine(`Using cached binary: ${binPath}`)
-      const valid = await new Promise<boolean>((resolve) => {
-        try {
-          execFile(binPath, ["version"], (err) => {
-            if (err) {
-              this.output.appendLine("Check for binary corruption: " + err)
-            }
-            resolve(err === null)
-          })
-        } catch (ex) {
-          this.output.appendLine("The cached binary cannot be executed: " + ex)
-          resolve(false)
-        }
-      })
-      if (valid) {
-        return binPath
-      }
-    }
+    const binPath = this.binaryPath()
+    const exists = await this.checkBinaryExists(binPath)
```

Comment:
Do we need to worry about error handling here?

Reply:
Good question! So this method calls `fs.stat` and has a `catch` method which returns `false` basically saying the path/file doesn't exist. I think it's okay as is for now!

Reply:
Oh ya! Looks great.